### DISCUSSION TITLE
Building on macOS Ventura

### DISCUSSION THREAD

**Discussion by jaypeeds** [ra3xdh/qucs_s#181] [Q&A]
Hello,
I have sorted out a few workarounds, none implying code change, to manage to get to some results.
1) Duplicated an existing icns file to provide a missing one:
```cp <git workspace>/qucs_s/qucs/bitmaps/qucsfilter.icns <git workspace>/qucs_s/qucs/bitmaps/qucspowercombining.icns```
2) Defined an environment variable to locate Homebrew installation of Qt5.
```export Qt5_DIR=/usr/local/Cellar/qt@5/5.15.6/lib/cmake/Qt5``` 
3) After the 'cmake ..'  step and before the 'make' step, had to rename qucs_s/VERSION file to qucs_s/VERSION.TXT, otherwise clang would resolve a #include \<version\> to this text file, obviously not in C++ syntax.
With these changes, 'make' would run with some warnings, and after a build without error, individual apps would run in their respective build directories.
But during the 'sudo make install' step, there is an error message for 3 out of the 8 components to be installed (-) indicates an error, (+) a success:
(+) qucs-activefilter
(+) qucs-filter
(+) qucs-powercombining
(+) library
(+) translations
(-) qucs
(-) qucs-attenuator
(-) qucs-transcalc

The error message is:
```CMake Error at qucs/cmake_install.cmake:45 (file):
  file INSTALL cannot find "//imageformats": No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:46 (include)
```
It looks like the location of Qt plugins directory is not resolved, but the installation by Homebrew does have a populated imageformats/ sub-directory.
And indeed, the 3 cmake_install.cmake files in the build/qucs, build/qucs-attenuator and build/qucs-transcalc sub-directory should be edited with:
``` Set(Qt5_DIR /usr/local/Cellar/qt@5/5.15.6)```

and the line with /imageformats edited to include the full path, for example:
```
if(CMAKE_INSTALL_COMPONENT STREQUAL "Runtime" OR NOT CMAKE_INSTALL_COMPONENT)
      file(INSTALL DESTINATION "${CMAKE_INSTALL_PREFIX}/bin/qucstrans.app/Contents/MacOS/plugins" TYPE DIRECTORY FILES "${Qt5_DIR}/plugins/imageformats")
  endif()
```

And those changes allow the 'sudo make install' step to complete, though with yet a last error: 
```
CMake Error at /usr/local/Cellar/cmake/3.24.2/share/cmake/Modules/BundleUtilities.cmake:997 (message):
  error: fixup_bundle: not a valid bundle
Call Stack (most recent call first):
  qucs-transcalc/cmake_install.cmake:60 (fixup_bundle)
  cmake_install.cmake:51 (include)
```

Hoping this can trigger some research by other devs... Could someone help ?
JP

**Comment by jaypeeds** [comment]
Answering to myself...
Instead of changing build files, now that the issue is understood, the fix should be added to qucs_s/\<failing component\>/CMakeLists.txt
Qt5_DIR and QT_PLUGINS_DIR should be set before 'cmake ..' step

Now, some additional findings:
The fixup_bundle error is caused by the translation of the APPS env variable:
For some reason, the app folder name in usr/local/bin is not matching the translation of the APPS variable:
$APPS: /usr/local/bin/qucsattenuator.app         and same for qucstrans.app
Actual: /usr/local/bin/qucs-sattenuator.app

However... Once the fixup_bundle is allowed to run with the proper bundle parameter value, the resulting bundle is not working, while it runs properly in its original form in the build directory. *fixup_bundle is corrupting the application bundle!*
To summarize:
1) Some of the automatically generated paths for the installed application location are wrong. As a result the app remains safe !
2) When they are correct, fixup_bundle can run but does a mess, the app is corrupted, and crashes at startup !

JP

**Comment by ra3xdh** [comment]
One of the previous Qucs developers had access to Apple hardware in the past and prepared a Mac package for Qucs. I have no access to MacOS and cannot debug the build issues on this platform. So, I dropped MacOS support for Qucs-S. I can only say that the reported CMake issues were never observed on LInux/WIndows(MXE,MSYS2). But I will certainly accept a patch providing a Mac compatibility if someone will submit it.

**Reply by jaypeeds** [reply]
Thank you Vadim, I can tell that it is possible to build the whole suite without any need of code change whatsoever. It builds with mild warnings like the usage of double-equal inside double-parentheses. All the tweaks are about installation, file layouts, paths and that sort of things, and it is just what it is since it was NOT part of your scope. Again, a big Thank You ! for this project !
JP

**Comment by 3j14** [comment]
In the past couple of days, I tired building for macOS again. Following @jaypeeds' instructions, I managed to compile the apps with Qt5 and Qt6. However, I had to make some changes to the `CMakeLists.txt` files to [build the applications statically](https://doc.qt.io/qt-6/macos-deployment.html#static-linking) (i.e. no local installation of Qt required).

I quickly want to document my steps so that other people can replicate it.

## Choice of package manager

While both [`brew`](https://brew.sh) and [`port`](https://www.macports.org) _should_ work, for now I only managed to compile a static binary using `port`. As the paths for both installations differ greatly, I will still point out the correct paths for both package managers. Maybe someone will get `brew` to work.

## Installing requirements

Depending on your package manager of choice, different package names are required. `bison` may not be required, though the version that ships with macOS (located at `/usr/bin/bison`) is outdated and did not work for me.

### Brew

```shell
brew install cmake bison qt6
```

### MacPorts

```shell
sudo port install cmake bison qt6-qtbase qt6-qtsvg qt6-qttools
```

Replace 'qt6' with 'qt5' if you prefer to use Qt 5 instead.

## Fix `CMakeLists.txt` files: Plugins

For static linking, [some Qt plugins need to be copied](https://doc.qt.io/qt-6/macos-deployment.html#qt-plugins) into the correct directory in the macOS app bundle. The current `CMakeList.txt` file only copies the 'imageformats' plugins into the `qucs-s.app/Contents/MacOS` directory. According to [Qt's documentation](https://doc.qt.io/qt-6/macos-deployment.html#qt-plugins), they should be placed in the `qucs-s.app/Contents/plugins` directory.

The following modifications have to be made to the `CMakeList.txt` files:

```cmake
# ...

IF(APPLE)
  # Replace this:
  SET(plugin_dest_dir ${PROJECT_NAME}.app/Contents/MacOS)
  # with this:
  SET(plugin_dest_dir ${PROJECT_NAME}.app/Contents)
  
  # ...
ENDIF(APPLE)

# ...

IF(APPLE AND QT_PLUGINS_DIR)
  INSTALL(DIRECTORY "${QT_PLUGINS_DIR}/imageformats" DESTINATION bin/${plugin_dest_dir}/plugins COMPONENT Runtime)
  # Add these two lines, adding 'platforms' and 'styles' plugins:
  INSTALL(DIRECTORY "${QT_PLUGINS_DIR}/platforms" DESTINATION bin/${plugin_dest_dir}/plugins COMPONENT Runtime)
  INSTALL(DIRECTORY "${QT_PLUGINS_DIR}/styles" DESTINATION bin/${plugin_dest_dir}/plugins COMPONENT Runtime)
ENDIF()
```

Make these changes to these three files:

- `qucs_s/qucs/CMakeLists.txt`
- `qucs_s/qucs-attenuator/CMakeLists.txt`
- `qucs_s/qucs-transcalc/CMakeLists.txt`

> **Warning**
> For `qucs-attenuator` and `qucs-transcalc`, you need to use `${QUCS_NAME}` instead of `${PROJECT_NAME}` and adjust the path according to the previous value.

## Building

```shell
mkdir build
cd build
cmake .. -DQT_DIR=<qt_cmake_dir> -DQt6_DIR=<qt_cmake_dir> -DCMAKE_INSTALL_PREFIX=../dist -DQT_PLUGINS_DIR=<plugins_dir>
mv ../VERSION ../VERSION.txt  # thanks @jaypeeds for pointing out this problem!
make  # or 'make -j9' to run with 9 jobs in parallel.
make install
```

Replace `<qt_cmake_dir>` and `<plugins_dir>` with the values taken from the table below:

| Package manager | Qt major version | `<qt_cmake_dir>` | `<plugins_dir>` |
| --------------- | ---------------- | ---------------- | --------------- |
| `port` | 5 | `/opt/local/lib/cmake/Qt5` | `/opt/local/libexec/qt5/plugins` |
| `port` | 6 | `/opt/local/libexec/qt6/lib/cmake/Qt6` | `/opt/local/libexec/qt6/plugins` |
| `brew` | 6 | `/opt/homebrew/lib/cmake/Qt6` | `/opt/homebrew/share/qt/plugins` |

> **Note**
> I didn't test Brew with Qt 5 because Qt 6 already caused some problems.

The binaries should finally be located in the `qucs_s/dist/bin` directory.

## Other remarks

I only tested this on macOS Ventura on Apple M2 (ARM). I can create a pull request with the changes (and maybe a better fix for the VERSION file) if there is interest in it. We could also setup a simple GitHub Action that automatically builds a macOS app bundle for release.

**Reply by 3j14** [reply]
For the sake of simplicity, [here is a patch](https://github.com/ra3xdh/qucs_s/files/12327360/macos-static-build.patch) that incorporates the changes made to the `CMakeLists.txt` files.

**Reply by nessex** [reply]
Thanks for all this! It worked for me. I needed to change one small thing though, the cmake command also required `-DWITH_QT6=ON`.

I also got a CMake error at the end of cmake install, but despite this the binary works 🤷‍♀️ , so I'm ignoring it until I find something broken. For now, circuits, simulations etc. seem fine!

<details>
    <summary>Error</summary>
    
```
CMake Error at /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/BundleUtilities.cmake:458 (message):
  otool -l failed: 1

  error:
  /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/otool-classic:
  can't open file: @rpath/QtCore.framework/Versions/A/QtCore (No such file or
  directory)

Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/BundleUtilities.cmake:527 (get_item_rpaths)
  /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/BundleUtilities.cmake:614 (set_bundle_key_values)
  /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/BundleUtilities.cmake:933 (get_bundle_keys)
  qucs/cmake_install.cmake:73 (fixup_bundle)
  cmake_install.cmake:46 (include)

```
</details>

Here's the full set of commands I used to get set up with qucs-s + ngspice on an M1 mac (with brew installed for ARM in /opt/homebrew, and using a terminal under arm not x86/rosetta):

```shell
# Install ngspice
brew install ngspice

# Install qucs-s
git clone https://github.com/ra3xdh/qucs_s.git
cd qucs_s
curl -fsSL -O https://github.com/ra3xdh/qucs_s/files/12327360/macos-static-build.patch
git apply ./macos-static-build.patch
mkdir build
cd build
cmake .. -DQT_DIR=/opt/homebrew/lib/cmake/Qt6 -DQt6_DIR=/opt/homebrew/lib/cmake/Qt6 -DCMAKE_INSTALL_PREFIX=../dist -DQT_PLUGINS_DIR=/opt/homebrew/share/qt/plugins -DWITH_QT6=ON
mv ../VERSION ../VERSION.txt
make -j9
make install

# Run qucs-s
../dist/bin/qucs-s.app/Contents/MacOS/qucs-s

# Initial configuration
ngspice location: /opt/homebrew/bin/ngspice
```

Naturally, `../dist/bin/qucs-s.app` could be moved to `Applications` or wherever to "install" it in a more typical fashion for MacOS as well.
